File: tests/node-compat/stage-571/child-process-events.js
const assert = require("assert");
const childProcess = require("child_process");

const child = childProcess.spawn("node", ["-e", ""]);
let spawned = false;
child.on("spawn", () => {
  spawned = true;
});
child.once("exit", (code, signal) => {
  assert.strictEqual(spawned, true);
  assert.strictEqual(code, 0);
  assert.strictEqual(signal, null);
});
child.once("close", (code, signal) => {
  assert.strictEqual(spawned, true);
  assert.strictEqual(code, 0);
  assert.strictEqual(signal, null);
  console.log("child process events passed");
});

File: crates/quench-node/src/polyfills/bootstrap/child_process_events.rs
//! Polyfill: `child-process-events`

pub const JS: &str = quench_js_check::checked_js!(r#"const __quenchChildEventOrderRequire = globalThis.require;
const __quenchChildEventOrder = __quenchChildEventOrderRequire("child_process");
const __quenchEventOrderSpawn = __quenchChildEventOrder.spawn;
__quenchChildEventOrder.spawn = (...args) => {
  const child = __quenchEventOrderSpawn(...args);
  child.on = globalThis.__nodeEventEmitter.prototype.on;
  child.emit = (
    (emit) => (event, ...values) => {
      if (event === "exit" && !child.__spawnEmitted) {
        child.__spawnEmitted = true;
        Reflect.apply(emit, child, ["spawn"]);
      }
      return Reflect.apply(emit, child, [event, ...values]);
    }
  )(child.emit);
  return child;
};
"#);
